Compile error with recent nightly toolchain

rustc version: `rustc 1.69.0-nightly (dc1d9d50f 2023-01-31)`

Reproduce: run `cargo test` at the root directory. (Maybe `rust-toolchain` should be deleted first.)
```
error[E0161]: cannot move a value of type `str`
  --> utils/zerovec/derive/examples/make_var.rs:10:1
   |
10 | #[make_varule(VarStructULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the size of `str` cannot be statically determined
   |
   = note: this error originates in the attribute macro `make_varule` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `self.c` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:10:1
   |
10 | #[make_varule(VarStructULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `self.c` has type `str`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `other.c` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:10:1
   |
10 | #[make_varule(VarStructULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `other.c` has type `str`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0161]: cannot move a value of type `str`
  --> utils/zerovec/derive/examples/make_var.rs:20:1
   |
20 | #[make_varule(VarStructOutOfOrderULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the size of `str` cannot be statically determined
   |
   = note: this error originates in the attribute macro `make_varule` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `self.b` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:20:1
   |
20 | #[make_varule(VarStructOutOfOrderULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `self.b` has type `str`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `other.b` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:20:1
   |
20 | #[make_varule(VarStructOutOfOrderULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `other.b` has type `str`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0161]: cannot move a value of type `VarZeroSlice<str>`
  --> utils/zerovec/derive/examples/make_var.rs:31:1
   |
31 | #[make_varule(VarTupleStructULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the size of `VarZeroSlice<str>` cannot be statically determined
   |
   = note: this error originates in the attribute macro `make_varule` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `self` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:31:1
   |
31 | #[make_varule(VarTupleStructULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `self.2` has type `VarZeroSlice<str>`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `other` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:31:1
   |
31 | #[make_varule(VarTupleStructULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `other.2` has type `VarZeroSlice<str>`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0161]: cannot move a value of type `VarZeroSlice<str>`
  --> utils/zerovec/derive/examples/make_var.rs:36:1
   |
36 | #[make_varule(NoKVULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^ the size of `VarZeroSlice<str>` cannot be statically determined
   |
   = note: this error originates in the attribute macro `make_varule` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `self` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:36:1
   |
36 | #[make_varule(NoKVULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `self.2` has type `VarZeroSlice<str>`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `other` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:36:1
   |
36 | #[make_varule(NoKVULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `other.2` has type `VarZeroSlice<str>`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0161]: cannot move a value of type `VarZeroSlice<str>`
  --> utils/zerovec/derive/examples/make_var.rs:42:1
   |
42 | #[make_varule(NoOrdULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^ the size of `VarZeroSlice<str>` cannot be statically determined
   |
   = note: this error originates in the attribute macro `make_varule` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `self` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:42:1
   |
42 | #[make_varule(NoOrdULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `self.2` has type `VarZeroSlice<str>`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `other` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:42:1
   |
42 | #[make_varule(NoOrdULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `other.2` has type `VarZeroSlice<str>`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0161]: cannot move a value of type `MultiFieldsULE`
  --> utils/zerovec/derive/examples/make_var.rs:48:1
   |
48 | #[make_varule(MultiFieldStructULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the size of `MultiFieldsULE` cannot be statically determined
   |
   = note: this error originates in the attribute macro `make_varule` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `self.unsized_fields` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:48:1
   |
48 | #[make_varule(MultiFieldStructULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `self.unsized_fields` has type `MultiFieldsULE`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `other.unsized_fields` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:48:1
   |
48 | #[make_varule(MultiFieldStructULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `other.unsized_fields` has type `MultiFieldsULE`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0161]: cannot move a value of type `MultiFieldsULE`
  --> utils/zerovec/derive/examples/make_var.rs:62:1
   |
62 | #[make_varule(MultiFieldTupleULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the size of `MultiFieldsULE` cannot be statically determined
   |
   = note: this error originates in the attribute macro `make_varule` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `self` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:62:1
   |
62 | #[make_varule(MultiFieldTupleULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `self.2` has type `MultiFieldsULE`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `other` which is behind a shared reference
  --> utils/zerovec/derive/examples/make_var.rs:62:1
   |
62 | #[make_varule(MultiFieldTupleULE)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `other.2` has type `MultiFieldsULE`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0161, E0507.
For more information about an error, try `rustc --explain E0161`.
error: could not compile `zerovec-derive` due to 21 previous errors
```

--- a/ffi/gn/icu4x/BUILD.gn
+++ b/ffi/gn/icu4x/BUILD.gn
@@ -1447,7 +1447,7 @@ rust_library("zerovec-v0_9_2") {
   deps = []
   deps += [ ":yoke-v0_7_0" ]
   deps += [ ":zerofrom-v0_1_1" ]
-  deps += [ ":zerovec-derive-v0_9_2($host_toolchain)" ]
+  deps += [ ":zerovec-derive-v0_9_3($host_toolchain)" ]
 
   rustenv = []
 
@@ -1463,10 +1463,10 @@ rust_library("zerovec-v0_9_2") {
   visibility = [ ":*" ]
 }
 
-rust_proc_macro("zerovec-derive-v0_9_2") {
+rust_proc_macro("zerovec-derive-v0_9_3") {
   crate_name = "zerovec_derive"
   crate_root = "//utils/zerovec/derive/src/lib.rs"
-  output_name = "zerovec_derive-7ef1f6966b3a74c7"
+  output_name = "zerovec_derive-4d58846cb210946f"
 
   deps = []
   deps += [ ":proc-macro2-v1_0_32" ]
@@ -1479,8 +1479,8 @@ rust_proc_macro("zerovec-derive-v0_9_2") {
   rustflags = [
     "--cap-lints=allow",
     "--edition=2018",
-    "-Cmetadata=7ef1f6966b3a74c7",
-    "-Cextra-filename=-7ef1f6966b3a74c7",
+    "-Cmetadata=4d58846cb210946f",
+    "-Cextra-filename=-4d58846cb210946f",
   ]
 
   visibility = [ ":*" ]

--- a/utils/zerovec/derive/Cargo.toml
+++ b/utils/zerovec/derive/Cargo.toml
@@ -4,7 +4,7 @@
 
 [package]
 name = "zerovec-derive"
-version = "0.9.2"
+version = "0.9.3"
 description = "Custom derive for the zerovec crate"
 repository = "https://github.com/unicode-org/icu4x"
 license = "Unicode-DFS-2016"

--- a/utils/zerovec/derive/src/make_varule.rs
+++ b/utils/zerovec/derive/src/make_varule.rs
@@ -115,7 +115,6 @@ pub fn make_varule_impl(attr: AttributeArgs, mut input: DeriveInput) -> TokenStr
     let doc = format!("[`VarULE`](zerovec::ule::VarULE) type for {name}");
     let varule_struct: DeriveInput = parse_quote!(
         #[repr(#repr_attr)]
-        #[derive(PartialEq, Eq)]
         #[doc = #doc]
         #vis struct #ule_name #field_inits #semi
     );
@@ -140,6 +139,19 @@ pub fn make_varule_impl(attr: AttributeArgs, mut input: DeriveInput) -> TokenStr
         &ule_name,
         lt,
         input.span(),
+    );
+
+    let eq_impl = quote!(
+        impl core::cmp::PartialEq for #ule_name {
+            fn eq(&self, other: &Self) -> bool {
+                // The VarULE invariants allow us to assume that equality is byte equality
+                // in non-safety-critical contexts
+                <Self as zerovec::ule::VarULE>::as_byte_slice(&self)
+                == <Self as zerovec::ule::VarULE>::as_byte_slice(&other)
+            }
+        }
+
+        impl core::cmp::Eq for #ule_name {}
     );
 
     let zerofrom_fq_path =
@@ -233,6 +245,8 @@ pub fn make_varule_impl(attr: AttributeArgs, mut input: DeriveInput) -> TokenStr
         #derived
 
         #maybe_ord_impls
+
+        #eq_impl
 
         #zmkv
 

